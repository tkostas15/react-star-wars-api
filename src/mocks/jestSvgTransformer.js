export default 'MockedSvg'
export const ReactComponent = (props) => {
    return <svg {...props}></svg>
}