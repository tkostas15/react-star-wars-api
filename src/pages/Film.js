// imports
import { useNavigate, useParams }        from "react-router-dom";
import FilmInfo                          from '../components/FilmInfo/FilmInfo';
import { Fragment, useEffect, useState } from "react";
import Spinner                           from "../components/UI/Spinner";
import loadSpinner                       from "../icons/darth_vader_loader.png";
import Header                            from "../components/Header/Header";
import Modal                             from "../components/UI/Modal";
import Footer                            from "../components/Footer/Footer";
import { useDispatch, useSelector }      from "react-redux";
import { fetchFilmInfoStart }            from '../store/filmSlice';

const Film = (props) => {
    // get router parameters
    const routerParams = useParams();
    
    // retrieve slice states
    const fetchedInfo        = useSelector((state) => state.filmReducer.fetchedInfo);
    const httpIsSending      = useSelector((state) => state.filmReducer.sending);
    const httpError          = useSelector((state) => state.filmReducer.error);
    const httpAlreadyFetched = useSelector((state) => state.filmReducer.atLeastOneFetch);
    
    // states
    const [retry, setRetry] = useState(false);
    
    // film id
    const filmId = routerParams.filmId;
    
    // history
    const history = useNavigate();
    
    // dispatch action
    const dispatchAction = useDispatch();
    
    // fetch movie's info
    useEffect(() => {
        const apiToUse = {api: 'https://swapi.dev/api/films/' + filmId + '/'};
        dispatchAction(fetchFilmInfoStart(apiToUse));
    }, [dispatchAction, retry, filmId]);
    
    // error values
    let errorDescription;
    if (httpError) {
        errorDescription = (httpError === 'abort') ?
                           <div>Spacecraft could not land on this inhabited land</div> :
                           <div>Error on landing!</div>;
    }
    
    // retry
    const retryHandler = () => {setRetry(true)};
    
    // back handler
    const goBackHandler = () => { history(-1); };
    
    // return info at last
    return <Fragment>
        <Header hasBack={1} onClickBack={goBackHandler} />
        <main>
            {!httpIsSending && !httpError && httpAlreadyFetched && <FilmInfo info={fetchedInfo} />}
            {!httpIsSending && httpError && <Modal description={errorDescription} button='retry' onClick={retryHandler} />}
            {httpIsSending && <Spinner icon={loadSpinner} ms="1500" />}
        </main>
        <Footer />
    </Fragment>;
};

export default Film;