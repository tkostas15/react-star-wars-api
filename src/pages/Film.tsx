// imports
import { useNavigate, useParams } from "react-router-dom";
import FilmInfo from "../components/FilmInfo/FilmInfo";
import { Fragment, useEffect, useState } from "react";
import Spinner from "../components/UI/Spinner";
import loadSpinner from "../icons/darth_vader_loader.png";
import Header from "../components/Header/Header";
import Modal from "../components/UI/Modal";
import Footer from "../components/Footer/Footer";
import { useDispatch, useSelector } from "react-redux";
import { fetchFilmInfoStart } from "../store/filmSlice";
import {
  getFetchedInfo,
  getIsSending,
  getError,
  getIfFetched,
} from "../store/filmSlice";
import { FilmObject } from "../types/AllTypes";

const Film = (props: any) => {
  // get router parameters
  const routerParams = useParams();

  // retrieve slice states
  const fetchedInfo: Array<FilmObject> = useSelector(getFetchedInfo);
  const httpIsSending: boolean = useSelector(getIsSending);
  const httpError: string = useSelector(getError);
  const httpAlreadyFetched: boolean = useSelector(getIfFetched);

  // states
  const [retry, setRetry] = useState(false);

  // film id
  const filmId = routerParams.filmId;

  // history
  const history = useNavigate();

  // dispatch action
  const dispatchAction = useDispatch();

  // fetch movie's info
  useEffect(() => {
    const apiToUse = { api: "https://swapi.dev/api/films/" + filmId + "/" };
    dispatchAction(fetchFilmInfoStart(apiToUse));
  }, [dispatchAction, retry, filmId]);

  // error messages
  const abortError = (
    <div>Spacecraft could not land on this inhabited land</div>
  );
  const anyError = <div>Error on landing!</div>;

  // error values
  let errorDescription;
  if (httpError) {
    errorDescription = httpError === "abort" ? abortError : anyError;
  }

  // retry
  const retryHandler = () => {
    setRetry(!retry);
  };

  // back handler
  const goBackHandler = () => {
    history(-1);
  };

  // return info at last
  return (
    <Fragment>
      <Header hasBack={1} onClickBack={goBackHandler} />
      <main>
        {!httpIsSending && !httpError && httpAlreadyFetched && (
          <FilmInfo info={fetchedInfo} />
        )}
        {!httpIsSending && httpError && (
          <Modal
            description={errorDescription}
            button="retry"
            onClick={retryHandler}
          />
        )}
        {httpIsSending && <Spinner icon={loadSpinner} ms="1500" />}
      </main>
      <Footer />
    </Fragment>
  );
};

export default Film;
