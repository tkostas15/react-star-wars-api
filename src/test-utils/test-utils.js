// test-utils.js
// https://testing-library.com/docs/react-testing-library/setup
import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import createSagaMiddleware from "redux-saga";
import rootSagas from "../sagas/rootSagas";
import { BrowserRouter } from "react-router-dom";

import { rootReducer } from "../store/store";

let storeRedux;

// configure middlewares
const sagaMiddleware = createSagaMiddleware();

function render(
  Component,
  {
    initialState = {},
    middlewares = [],
    debug = false,
    extraProps,
    sagas = rootSagas,
    ...renderOptions
  } = {}
) {
  const store = configureStore({
    reducer: rootReducer,
    middleware: [sagaMiddleware],
    preloadedState: initialState,
  });
  function Wrapper({ children, ...restProps }) {
    return (
      <BrowserRouter>
        <Provider store={store} {...restProps}>
          {children}
        </Provider>
      </BrowserRouter>
    );
  }

  storeRedux = store;

  if (sagas instanceof Function) {
    if (debug) console.log("running sagas", sagas);
    sagaMiddleware.run(sagas);
  }

  return rtlRender(<Component {...store} {...extraProps} {...initialState} />, {
    wrapper: Wrapper,
    ...renderOptions,
  });
}

// re-export everything
export * from "@testing-library/react";

// override render method
export { render, storeRedux };
