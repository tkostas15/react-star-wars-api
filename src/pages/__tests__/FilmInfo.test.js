// import component
import Film from "../Film";

// import custom utils
import {render, screen} from "../../test-utils/test-utils";

// import msw
import {mswServer} from "../../mocks/server";
import {mockFilmsResponseHandlers} from "./exclude/mockFilmsResponseHandlers";

// mock router params
jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useParams: () => ({
        filmId: 1,
    }),
}));

// [TEST SUITE]
describe("[TEST SUITE] Check film info page", () => {
    beforeEach(() => {
        mswServer.use(...mockFilmsResponseHandlers);

        const FilmComponent = () => <Film/>;
        render(FilmComponent);
    });

    it("[TEST] Check film info table exists", async () => {
        const table = await screen.findByTestId("film");
        expect(table).toMatchSnapshot();
        expect(table).toBeInTheDocument();
    });

    it("[TEST] Check table only has one row", async () => {
        const $ = require("jquery");
        const table = await screen.findByTestId("film");
        expect($("tbody tr")).toHaveLength(1);
    });
});
