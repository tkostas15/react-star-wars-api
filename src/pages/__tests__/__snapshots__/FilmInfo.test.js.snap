// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[TEST SUITE] Check film info page [TEST] Check film info table exists 1`] = `
<div
  class="tableWrapper"
  data-testid="film"
>
  <div
    class="react-bootstrap-table"
  >
    <table
      class="table table-bordered table"
    >
      <thead>
        <tr>
          <th
            tabindex="0"
          >
            Title
          </th>
          <th
            tabindex="0"
          >
            Opening crawl
          </th>
          <th
            tabindex="0"
          >
            Director
          </th>
          <th
            tabindex="0"
          >
            Producer(s)
          </th>
          <th
            tabindex="0"
          >
            Release date
          </th>
          <th
            tabindex="0"
          >
            Characters (No)
          </th>
          <th
            tabindex="0"
          >
            Planets (No)
          </th>
          <th
            tabindex="0"
          >
            Starships (No)
          </th>
          <th
            tabindex="0"
          >
            Vehicles (No)
          </th>
          <th
            tabindex="0"
          >
            Species (No)
          </th>
          <th
            tabindex="0"
          >
            Created date
          </th>
          <th
            tabindex="0"
          >
            Edited date
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            style="white-space: nowrap; line-height: 1.5;"
          >
            A New Hope
          </td>
          <td
            style="font-style: italic; line-height: 1.5; min-width: 450px;"
          >
            It is a period of civil war.
Rebel spaceships, striking
from a hidden base, have won
their first victory against
the evil Galactic Empire.

During the battle, Rebel
spies managed to steal secret
plans to the Empire's
ultimate weapon, the DEATH
STAR, an armored space
station with enough power
to destroy an entire planet.

Pursued by the Empire's
sinister agents, Princess
Leia races home aboard her
starship, custodian of the
stolen plans that can save her
people and restore
freedom to the galaxy....
          </td>
          <td
            style="white-space: nowrap; line-height: 1.5;"
          >
            George Lucas
          </td>
          <td
            style="white-space: nowrap; line-height: 1.5;"
          >
            <div>
              Gary Kurtz
            </div>
            <div>
              Rick McCallum
            </div>
          </td>
          <td
            style="white-space: nowrap; line-height: 1.5;"
          >
            <div>
              May 25th, 1977
            </div>
          </td>
          <td
            style="white-space: nowrap; line-height: 1.5;"
          >
            18
          </td>
          <td
            style="white-space: nowrap; line-height: 1.5;"
          >
            3
          </td>
          <td
            style="white-space: nowrap; line-height: 1.5;"
          >
            8
          </td>
          <td
            style="white-space: nowrap; line-height: 1.5;"
          >
            4
          </td>
          <td
            style="white-space: nowrap; line-height: 1.5;"
          >
            5
          </td>
          <td
            style="white-space: nowrap; line-height: 1.5;"
          >
            <div>
              Dec 10th, 2014
            </div>
          </td>
          <td
            style="white-space: nowrap; line-height: 1.5;"
          >
            <div>
              Dec 20th, 2014
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
